wallet/wallet.go

```diff
@@ -230,6 +237,105 @@ func (w *SingleAddressWallet) SignTransaction(cs consensus.State, txn *types.Tra
 	return nil
 }
 
+// Split returns a transaction that splits the wallet in the given number of
+// outputs with given amount. It also returns a list of output IDs that need to
+// be signed.
+//
+// NOTE: split needs to use a minimal set of inputs and therefore does not reuse
+// the fund logic which randomizes the unspent transaction outputs used to fund
+// the transaction
+func (w *SingleAddressWallet) Split(cs consensus.State, outputs int, amount, feePerByte types.Currency, pool []types.Transaction) (types.Transaction, []types.OutputID, error) {
+	// prepare all outputs
+	var txn types.Transaction
+	for i := 0; i < int(outputs); i++ {
+		txn.SiacoinOutputs = append(txn.SiacoinOutputs, types.SiacoinOutput{
+			Value:      amount,
+			UnlockHash: w.Address(),
+		})
+	}
+
+	// fetch unspent transaction outputs
+	utxos, err := w.store.UnspentSiacoinElements()
+	if err != nil {
+		return types.Transaction{}, nil, err
+	}
+
+	// desc sort
+	sort.Slice(utxos, func(i, j int) bool {
+		return utxos[i].Value.Cmp(utxos[j].Value) > 0
+	})
+
+	// map used outputs
+	inPool := make(map[types.OutputID]bool)
+	for _, ptxn := range pool {
+		for _, in := range ptxn.SiacoinInputs {
+			inPool[types.OutputID(in.ParentID)] = true
+		}
+	}
+
+	// estimate the fees
+	outputFees := feePerByte.Mul64(uint64(len(encoding.Marshal(txn.SiacoinOutputs))))
+	feePerInput := feePerByte.Mul64(BytesPerInput)
+
+	// collect outputs that cover the total amount
+	var inputs []SiacoinElement
+	want := amount.Mul64(uint64(outputs))
+	for _, sce := range utxos {
+		inUse := w.used[sce.ID] || inPool[sce.ID]
+		matured := cs.Index.Height >= sce.MaturityHeight
+		sameValue := sce.Value.Equals(amount)
+		if inUse || sameValue || !matured {
+			continue
+		}
+
+		inputs = append(inputs, sce)
+		fee := feePerInput.Mul64(uint64(len(inputs))).Add(outputFees)
+		if SumOutputs(inputs).Cmp(want.Add(fee)) > 0 {
+			break
```

Comment:
(nit)

I feel like this doesn't quite match the description. Instead of splitting the wallet in the given number of outputs with the given amount, we seem to just try to guarantee that there are at least `n` outputs with a given value in the wallet. The actual number of outputs can both increase and decrease as part of this operation.

So maybe you can update the description to better reflect that.

Reply:
Perhaps `Redistribute` instead of `Split`?

Reply:
I guess we should revisit the original purpose of this functionality.

The motivating example is that we have a newly-funded wallet, with 1 big UTXO, and we want to form a bunch of contracts. Splitting our 1 big UXTO into e.g. 50 smaller UTXOs will allow us to form those contracts without creating a huge txn dependency chain.

Now consider a wallet that has 1000 tiny UTXOs. Is there any benefit to consolidating these into 50 larger UTXOs? Not really. We'll still be able to form contracts just fine. So we shouldn't be calling `Split` on a wallet that already has lots of UTXOs.

What we want is for the autopilot to have a fairly dumb loop like:
```go
for {
    if numUTXOs() < minUTXOs {
        makeMoreUTXOs()
    }
}
```
`makeMoreUTXOs` could either be an explicit bus endpoint (like `Split`/`Redistribute`), or it could be built out of lower-level endpoints like `/wallet/fund`. I'm slightly in favor of the former, mostly to make things easier for other autopilots.

Reply:
I like `Redistribute` 👍 - we should update `walletd` as well (I can do that if this PR gets approved).